include/pgduckdb/catalog/pgduckdb_catalog.hpp

```diff
@@ -0,0 +1,55 @@
+#pragma once
```

Comment:
@bleskes was wondering what the benefits are of making this change. He thought you probably had some good reasons, but that it would be good to call those out clearly in the PR description.

Afaict one reason is to be able to support inserts using the custom INSERT. Are there more?

If it's only inserts, another approach to be to split inserts into two:

1. A read-only part, which is executed by DuckDB
2. A writing part that actually writes into the table, which takes the output from the read-only part

Reply:
Anything beyond a simple replacement scan will need this, that's why we're opting to add it sooner rather than later.
I already need it for the enum type support:

```
(lldb) p query 
(const std::string &) "SELECT a FROM tbl WHERE (a = 'sad'::mood)"
```

The produced query from postgres contains `::mood` which means it is going to do a lookup in the catalog for the `mood` type, but that doesn't exist

Reply:
> If it's only inserts, another approach for supporting inserts would to be to split inserts into two:

A read-only part, which is executed by DuckDB, this generates the data that needs to be inserted
A writing part that actually writes into the table, which is executed Postgres and takes the output from the read-only part as its input

Isn't that exactly what this facilitates?
By overloading `PlanInsert` we can create a PhysicalOperator that performs the insert into postgres, being fed data from the duckdb plan

This is `DuckCatalog::PlanInsert`
```c++
unique_ptr<PhysicalOperator> DuckCatalog::PlanInsert(ClientContext &context, LogicalInsert &op,
                                                     unique_ptr<PhysicalOperator> plan) {
	bool parallel_streaming_insert = !PhysicalPlanGenerator::PreserveInsertionOrder(context, *plan);
	bool use_batch_index = PhysicalPlanGenerator::UseBatchIndex(context, *plan);
	auto num_threads = TaskScheduler::GetScheduler(context).NumberOfThreads();
	if (op.return_chunk) {
		// not supported for RETURNING (yet?)
		parallel_streaming_insert = false;
		use_batch_index = false;
	}
	if (op.action_type != OnConflictAction::THROW) {
		// We don't support ON CONFLICT clause in batch insertion operation currently
		use_batch_index = false;
	}
	if (op.action_type == OnConflictAction::UPDATE) {
		// When we potentially need to perform updates, we have to check that row is not updated twice
		// that currently needs to be done for every chunk, which would add a huge bottleneck to parallelized insertion
		parallel_streaming_insert = false;
	}
	unique_ptr<PhysicalOperator> insert;
	if (use_batch_index && !parallel_streaming_insert) {
		insert = make_uniq<PhysicalBatchInsert>(op.types, op.table, op.column_index_map, std::move(op.bound_defaults),
		                                        std::move(op.bound_constraints), op.estimated_cardinality);
	} else {
		insert = make_uniq<PhysicalInsert>(
		    op.types, op.table, op.column_index_map, std::move(op.bound_defaults), std::move(op.bound_constraints),
		    std::move(op.expressions), std::move(op.set_columns), std::move(op.set_types), op.estimated_cardinality,
		    op.return_chunk, parallel_streaming_insert && num_threads > 1, op.action_type,
		    std::move(op.on_conflict_condition), std::move(op.do_update_condition), std::move(op.on_conflict_filter),
		    std::move(op.columns_to_fetch));
	}
	D_ASSERT(plan);
	insert->children.push_back(std::move(plan));
	return insert;
}
```

Reply:
> The produced query from postgres contains `::mood` which means it is going to do a lookup in the catalog for the `mood` type, but that doesn't exist

That's a good point. I'm wondering whether we can get around this by simply removing `::mood` from the query that is generated by postgres by changing `duckdb_pg_get_querydef` to not include that cast. Duckdb does not actually seem to need it when I try some queries locally. But I might be missing some more complex queries where it actually requires the cast to be present.

> Isn't that exactly what this facilitates?

What I meant is that we don't necessarily need to send the `INSERT` to duckdb. So if a user runs:

```
INSERT INTO t SELECT * FROM read_csv(...);
```

Then we instead send `SELECT * FROM read_csv(...)` to duckdb, and let postgres logic insert the resulting tuples into table `t`.

Reply:
Yea that could work, using a streaming result we could even start inserting while the select is still executing

But like I said INSERT is definitely not the only reason we need this

Reply:
Yeah makes sense. For completeness, I found a type of query where where the casts definitely matter, and we cannot just leave them out:

```sql
D CREATE TYPE numbers AS ENUM ('3', '2', '1');
D select '1' < '3';
┌─────────────┐
│ ('1' < '3') │
│   boolean   │
├─────────────┤
│ true        │
└─────────────┘
D select '1'::numbers < '3'::numbers;
┌───────────────────────────────────────────────┐
│ (CAST('1' AS numbers) < CAST('3' AS numbers)) │
│                    boolean                    │
├───────────────────────────────────────────────┤
│ false                                         │
└───────────────────────────────────────────────┘
```

Reply:
Just to add a little more context:
`SELECT a FROM tbl WHERE (a = 'sad'::mood)` was produced from the query string:
`SELECT a FROM tbl WHERE a = 'sad'`

This works without the explicit cast because comparison between string and enum is not allowed, so it does a cast from the string to the enum it's being compared to